FHIR Server Configuration

nickclyde opened the pull request:
# PULL REQUEST

## Summary

Adds a page at `/fhir-servers` to add and edit FHIR servers. Based on [these designs](https://www.figma.com/design/Iuw9me6kYftBF4WTCEsCZz/Query-Connector?node-id=1155-4577&m=dev).

Currently resolves the following tickets:
https://linear.app/skylight-cdc/issue/QUE-98/create-basic-front-end-for-fhir-server-config
https://linear.app/skylight-cdc/issue/QUE-100/add-edit-and-delete-buttons-to-front-end-fhir-server-table
https://linear.app/skylight-cdc/issue/QUE-101/create-fhir-server-front-end-popup-for-create-new
https://linear.app/skylight-cdc/issue/QUE-102/write-a-creditedit-fhir-server-config-function
https://linear.app/skylight-cdc/issue/QUE-103/write-a-deletefhirserver-function-to-remove-a-db-config
https://linear.app/skylight-cdc/issue/QUE-104/finish-migrating-hard-coded-fhir-servers
https://linear.app/skylight-cdc/issue/QUE-113/update-fhirservers-and-fhir-servers

Screenshots:


<img width="1936" alt="Screenshot 2025-01-09 at 12 29 05 PM" src="https://github.com/user-attachments/assets/40f7d21a-241b-45ec-abac-787c41cc407a" />
<img width="1936" alt="Screenshot 2025-01-09 at 12 28 32 PM" src="https://github.com/user-attachments/assets/7c7c464d-9ca8-4228-a65a-47c7330a7887" />
<img width="1936" alt="Screenshot 2025-01-09 at 12 28 42 PM" src="https://github.com/user-attachments/assets/95c8d86a-d052-4dfd-bad2-781549f52ae9" />
<img width="1936" alt="Screenshot 2025-01-09 at 12 29 20 PM" src="https://github.com/user-attachments/assets/1d28c685-18a6-4ecb-a3ff-557885d1fa25" />
<img width="1936" alt="Screenshot 2025-01-09 at 12 29 36 PM" src="https://github.com/user-attachments/assets/ac33d738-b86d-4513-8119-3db52336dbf1" />

katyasoup commented on query-connector/src/app/fhir-servers/fhirServers.module.scss:
```diff
@@ -0,0 +1,62 @@
+
+  &:hover,
+  &:hover > td {
+    background-color: rgba(217, 232, 246, 0.50);
```
you should be able to apply the background to just the `&:hover` parent, and then the `td` children will inherit it (see [here](https://stackoverflow.com/questions/27332016/how-do-i-layer-css-opacity) for more of an explanation)

that seems to be working when I adjust it on my local, but I know I ran into a similar issue before where the child elements weren't picking it up, and I had to do some math-y magic (like, parent has 0.5, child has 0.125) - lmk if the above doesn't do the trick!

nickclyde replied:
Amazing, this worked!! Thanks @katyasoup!!

katyasoup reviewed (commented):
Looking good! Couple design things I noticed (ignore if these are out of scope!)

<img width="400" alt="Screenshot 2025-01-08 at 14 18 48" src="https://github.com/user-attachments/assets/49301ac3-ee57-4afe-839d-74a48407a957" />

- the Delete button looks like it should be left-justified per the Figma mockup

<img width="400" alt="Screenshot 2025-01-08 at 14 18 55" src="https://github.com/user-attachments/assets/81333ff8-ba50-4707-9573-82027191b3cb" />

- when I click Test Connection and get a success result, the buttons all grow in height; looks like Figma only shows the success response without the Delete button so I'd double check with @mikang on that
- "Save Changes" always results in a "Failed to update the server configuration." error (but maybe that is out of scope?)

mikang commented:
Thanks @katyasoup !

- The cancel button should be just a link, not a white button. Figma below:
<img width="498" alt="image" src="https://github.com/user-attachments/assets/79eae2e8-64c4-44b3-8b6b-1a1a1028e7ec" />

- Actually, that's a good catch. The Delete button should be there if the user is editing and then testing the connection. Can we keep the Delete button and change the text to "[checkmark] Success!" ? That shortens the text, so the buttons don't grow in height.

m-goggins commented:
Would you be able to add the tickets that this PR covers? It looks like a lot of the ones written up in December but it'd be helpful to see what is left before the Connectathon.

nickclyde commented:
@katyasoup @mikang Just pushed the design fixes for the stuff you pointed out!

@m-goggins added the list of tickets this addresses so far. I plan on implementing delete functionality next and moving the hardcoded servers into the table after that as discussed at standup!

robertandremitchell reviewed (commented):
looks good in `npm run dev`! I was able to successfully add and remove a fhir server without issue.

running into this error on `npm run build`

```

./src/app/fhir-servers.ts:14:3
Type error: Type '{ id: string; name: string; hostname: string; headers: {}; }' is not assignable to type 'DevFhirServerConfig'.
  Type '{ id: string; name: string; hostname: string; headers: {}; }' is missing the following properties from type 'FhirServerConfig': last_connection_attempt, last_connection_successful

  12 |   process.env.E2E_LOCAL_FHIR_SERVER ?? "http://hapi-fhir-server:8080/fhir";
  13 | export const fhirServers: Record<string, DevFhirServerConfig> = {
> 14 |   "HELIOS Meld: Direct": {
     |   ^
  15 |     id: "HELIOS Meld: Direct",
  16 |     name: "HELIOS Meld: Direct",
  17 |     hostname: "https://gw.interop.community/HeliosConnectathonSa/open",
```

fzhao99 reviewed (approved):
this lgtm! think the build error is a pretty simple fix, so approving.

Wondering if we should have a followup that adds our hardcoded default servers to the dump file? Might be good to have them available out of the box like we do the default queries.

m-goggins commented on query-connector/src/app/fhir-servers/page.tsx:
```diff
@@ -0,0 +1,392 @@
+   * @param url - The URL of the FHIR server to test
+   * @returns A promise that resolves to an object containing the test result
+   */
+  const testFhirConnection = async (
```
Definitely could be a separate ticket/post-connectathon but I think we should try to test the Connection upon insertion, at least for the default servers, so that when a user first looks at the table the status is not null.

nickclyde replied:
It actually already does this! See line 178

m-goggins reviewed (approved):
Looks good to me as long as we update the table to pull the list of available FHIR servers from the table and the hard coded list like we do for the drop down or update the table to include all of the hard coded servers (even is just for the connectathon). Left a few other comments for consideration but nothing that should be blocking.

katyasoup commented on query-connector/src/app/fhir-servers/page.tsx:
```diff
@@ -0,0 +1,392 @@
+      buttons[1].text = (
+        <>
+          <Icon.Check
+            size={3}
```
I'm still seeing the button size jump after the connection check, which I think is caused by the checkmark icon? Bumping this down to size 2 keeps it in line with the line height of the text

katyasoup replied:
(or decrease the top/bottom padding to accommodate, since I see the linter yelling about 2 not being an acceptable value here)

katyasoup commented on query-connector/src/app/fhir-servers/page.tsx:
```diff
@@ -0,0 +1,392 @@
+        </>
+      );
+      buttons[1].className =
+        "usa-button usa-button--outline shadow-none text-green padding-left-0 padding-right-2";
```
can we add the same left padding here so the checkmark + text are centered evenly?

nickclyde commented:
Just fixed the build by moving all the hardcoded servers into the database via migration. Also cleaned up the `FhirServers` constant references. Also had to make the import of `Modal` a dynamic import to make the build work. Still have a few comments to address but I think this now covers almost all of the FHIR server configuration tickets in Linear!

robertandremitchell reviewed (approved):
Looks good! a few formatting touch-ups, but I think they would be fine to tackle in a follow-up ticket if we want to get the bulk of this merged.

<img width="539" alt="Screenshot 2025-01-13 at 8 28 30 AM" src="https://github.com/user-attachments/assets/e2fb4f7c-bf3c-4658-8e10-581f77b77891" />

https://github.com/user-attachments/assets/c06e6fdb-746c-41e1-9ec6-ca5bf813ca3a

nickclyde commented on query-connector/src/app/fhir-servers/page.tsx:
```diff
@@ -0,0 +1,392 @@
+            break;
+          case 403:
+            errorMessage =
+              "Connection failed: Access forbidden. You do not have permission to access this FHIR server.";
+            break;
```
Hmm, I'm not sure. We can always adjust later, but I think since everything else on the page calls it FHIR server let's just leave it for now

nickclyde commented:
Good catch @robertandremitchell! Fixed those issues!
